### PR TITLE
 feat: Implement subset of crypto.subtle.importKey which can import a JSONWebKey using RSASSA-PKCS1-v1_5

### PR DESCRIPTION
**Pull request by JakeChampion**

### CODE REVIEW

File: c-dependencies/js-compute-runtime/builtins/json-web-key.cpp [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+
+namespace {
+JS::Result<std::optional<std::string>>
+extractStringPropertyFromObject(JSContext *cx, JS::HandleObject object, std::string_view property) {
+  bool has_property;
+  if (!JS_HasProperty(cx, object, property.data(), &has_property)) {
+    return JS::Result<std::optional<std::string>>(JS::Error());
+  }
+  if (!has_property) {
+    return std::optional<std::string>(std::nullopt);
```
Comment (elliottt) [inline c-dependencies/js-compute-runtime/builtins/json-web-key.cpp:28]: Does this change work, or did you run into template argument deduction problems?
```suggestion
    return std::nullopt;
```

File: c-dependencies/js-compute-runtime/builtins/json-web-key.cpp [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+          }
+
+          // No duplicates allowed
+          if (std::find(key_ops.begin(), key_ops.end(), op) != key_ops.end()) {
+            JS_ReportErrorASCII(
+                cx, "The 'key_ops' member of the JWK dictionary contains duplicate usages");
+            return nullptr;
+          }
+
+          key_ops.push_back(op);
```
Comment (elliottt) [inline c-dependencies/js-compute-runtime/builtins/json-web-key.cpp:250]: Using `push_back` here will copy the string, but `emplace_back` will give you the opportunity to invoke the move constructor for the target cell in the vector.
```suggestion
          key_ops.emplace_back(std::move(op));
```

File: c-dependencies/js-compute-runtime/builtins/crypto-algorithm.cpp [hunk 3]
```diff
@@ -183,6 +440,300 @@ std::unique_ptr<CryptoAlgorithmDigest> CryptoAlgorithmDigest::normalize(JSContex
...
+    // contains an entry which is not "verify" then throw a SyntaxError.
+    bool isUsagesAllowed = false;
+    // public key
+    if (jwk->d.has_value()) {
+      isUsagesAllowed = usages.canSign();
+    } else {
+      // private key
+      isUsagesAllowed = usages.canVerify();
+    }
+    isUsagesAllowed = isUsagesAllowed || !usages.isEmpty();
```
Comment (elliottt) [inline c-dependencies/js-compute-runtime/builtins/crypto-algorithm.cpp:541]: Why would usages be allowed if `usages` was not empty? The way I'm reading the comment it seems like the two cases are just those of the conditional above.

**Review by elliottt — COMMENTED** [review]
This is looking great!

File: c-dependencies/js-compute-runtime/builtins/json-web-key.cpp (hunk 1 above)
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/builtins/json-web-key.cpp:28]: This runs into the deduction problem you've mentioned -- 

>  no viable conversion from returned value of type 'const std::nullopt_t' to function return type 'JS::Result<std::optional<std::string>>

File: c-dependencies/js-compute-runtime/builtins/json-web-key.cpp (hunk 2 above)
Comment (elliottt) [reply c-dependencies/js-compute-runtime/builtins/json-web-key.cpp:250]: I partially take this back: there's no need to use `emplace_back` here, `push_back(std::move(op))` will get the same effect :+1:

**Review by elliottt — APPROVED** [review]
This looks great! Just a few suggestions before merging :+1:

File: c-dependencies/js-compute-runtime/builtins/crypto-key.cpp [hunk 4]
```diff
@@ -234,4 +328,194 @@ bool CryptoKey::init_class(JSContext *cx, JS::HandleObject global) {
...
+    if (keyData->privateExponent.empty() || keyData->firstPrimeInfo->primeFactor.empty() ||
+        keyData->secondPrimeInfo->primeFactor.empty()) {
+      return nullptr;
+    }
+  }
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+  auto rsa = RSA_new();
+#pragma clang diagnostic pop
```
[the remark this answers is not in the document]
Comment (JakeChampion) [reply c-dependencies/js-compute-runtime/builtins/crypto-key.cpp:386]: Issue at https://github.com/fastly/js-compute-runtime/issues/477

File: c-dependencies/js-compute-runtime/builtins/crypto-algorithm.cpp [hunk 5]
```diff
@@ -3,11 +3,258 @@
...
+    JS_ReportErrorLatin1(cx,
+                         "The JWK member 'n' could not be base64url decoded or contained padding");
+    return nullptr;
+  }
+  auto modulus = modulusResult.unwrap();
+  // Per RFC 7518 Section 6.3.1.1: https://tools.ietf.org/html/rfc7518#section-6.3.1.1
+  if (modulus.starts_with('0')) {
+    modulus = modulus.erase(0, 1);
+  }
+  auto dataResult = GlobalProperties::ConvertJSValueToByteString(cx, jwk->e.value());
```
Comment (JakeChampion) [inline c-dependencies/js-compute-runtime/builtins/crypto-algorithm.cpp:77]:
```suggestion
  auto dataResult = GlobalProperties::convertJSValueToByteString(cx, jwk->e.value());
```

**Comment by stefan-guggisberg** [comment]
@JakeChampion Thanks for landing this PR in 1.7.0.

I was hoping this would solve our problem (3rd party dependencies relying on `subtle.crypto` e.g. for signing AWS requests. However, we're now seeing different errors, such as `Supplied algorithm is not yet supported`.

Here's a code fragment from https://github.com/mhart/aws4fetch which seems to trigger this error:

```javascript
async function hmac(key2, string3) {
  const cryptoKey = await crypto.subtle.importKey(
    "raw",
    typeof key2 === "string" ? encoder2.encode(key2) : key2,
    { name: "HMAC", hash: { name: "SHA-256" } },
    false,
    ["sign"]
  );
  return crypto.subtle.sign("HMAC", cryptoKey, encoder2.encode(string3));
}
```

It seems like the lack of `HMAC` support is causing the error. 

Would this PR solve our problem? https://github.com/fastly/js-compute-runtime/pull/440

Can you give us an ETA when this PR lands in production?

**Comment by JakeChampion** [comment]
@stefan-guggisberg - #440 does solve that issue 👍  - I am breaking that PR up into smaller more manageable PR for the team to review. If you subscribe to #440, when it is closed you should be able to run the code you've posted without any issues 👍

**Pull request merged**